text/0032-upgrade-bootstrapper.md

```diff
@@ -0,0 +1,152 @@
+user _copies_ the installed vanilla config rather than renaming or deleting it.  Two possible solutions are:
+1. have the installer create a read-only copy of the default config exclusively for the `upgrade-bootstrapper`'s use,
+   and a separate writable one for the user to modify/copy/move as they wish
+1. require that every config value is annotated with a standard comment specifying the default value
+
+It seems like annotating the config file with default values might be somewhat more "normal" and should be fairly simple
+to implement and test.
+
+To migrate, the node will identify any modified values in the old config, and apply the same changes to the new config
+if the corresponding new config value has not already been modified by the user.
+
+
+### `upgrade-bootstrapper`
+
+We also need to provide an `upgrade-bootstrapper` binary.  It should live in its own repo outside the `casper-node` one
+if setting up CI is not deemed too onerous.  Ideally its tests shouldn't need to run when changes are made to
+`casper-node`, and vice-versa.  There also should be no direct dependency between the `update-bootstrapper` and any of
+the other casper crates.  Hence it makes sense to keep this separate to the rest of the node project.
+
+The `upgrade-bootstrapper` and `casper-node` must both be capable of being restarted, having been stopped or killed at
+any step above, without disrupting this normal flow.  This will require the `upgrade-bootstrapper` to write which step
+its about to perform to the hard-coded known location immediately before performing the step.  On starting, the
+`upgrade-bootstrapper` will try to read this value and jump to that step.  If the value/file is absent, it assumes it's
+a fresh start.
+
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+Users lose the ability to override config values via command line.
+
+
+## Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+1. How to get the upgrade activation point agreed by all validators?
+    * Could include first activation point in chainspec, and all chainspec upgrade files hold the _next_ activation
+      point. (Drawback is that we need to know much further in advance when we want the next activation point to occur)
+    * Once new chainspec upgrade file is installed, `upgrade-bootstrapper` could notify the node about its path (via
+      IPC or TCP listener on loopback).  Nodes could then use consensus to agree the point (could take a long time to
+      get consensus as install times of next version of software could vary widely)
+    * On each era change, `casper-node` could read from a file in a known location the next upgrade activation point.
+      This could be the chainspec upgrade file, or a standalone one containing just the activation point.
+
+1. Do we have to keep all chainspec upgrade files, or just the latest?  If just the latest, change the chainspec to not
```

Comment:
I think we need to. If consensus needs `UNBONDING_DELAY` eras worth of information to properly validate consensus messages, and there will be upgrades within last `UNBONDING_DELAY` eras, each era may have different protocol configuration that defines whether a messages is valid or not. cc @afck

Reply:
I guess so… except if we say that cross-era slashing doesn't work across upgrade boundaries; maybe that's the easier option?

Reply:
It seems so but is this a correct thing to do?

Reply:
Not quite, no. It does create a short window for unslashed equivocations. :confused:

Reply:
It's not only about equivocations though – every running node should gossip (and validate) consensus messages from the past eras. If we do not store old chainspecs we basically say that after protocol upgrade we do not accept any old message. We're "cutting off" all the past eras effectively.

Reply:
So, this proposal does really expect that nodes will not be able to communicate across protocol upgrades (at least major version change ones).  That's the main rationale for the proposal - to avoid the need for complex branching code which supports communications and data types which are incompatible.

I'd anticipated that a node would be able to communicate about previous eras using migrated data and using the new version of the message types.  For example, if we change the block structure in a version upgrade, migration would change old blocks to the new format, allowing them to still be gossiped.

Would that suffice for consensus, or do we _really_ need to support handling old versions of data and/or messages after a major protocol upgrade?

Reply:
It depends if we are OK sacrificing the some security of the protocol. Like @afck , in this case we open up ourselves to an attack where equivocators can continue unslashed. Personally, I think it's a bad idea and we should not do it.

Reply:
> So, this proposal does really expect that nodes will not be able to communicate across protocol upgrades (at least major version change ones). 

Sounds more like a solution for an emergency restart of the network and not a protocol upgrade.